### PR TITLE
Fix panic in `htmlimageelement.rs` using `char_indices()` to iterate over characters

### LINKED ISSUE #32976: Panic parsing srcset in html/semantics/embedded-content/the-img-element/srcset/parse-a-srcset-attribute.html
To reproduce: `./mach test-wpt tests/wpt/tests/html/semantics/embedded-content/the-img-element/srcset/parse-a-srcset-attribute.html`

The code in https://github.com/servo/servo/blob/main/components/script/dom/htmlimageelement.rs#L1894-L1897 is misusing the chars iterator to obtain byte indexes from the string. We should use the [char_indices](https://doc.rust-lang.org/stable/std/primitive.str.html#method.char_indices) iterator instead.

Backtrace:
```
pid:15614 byte index 1 is not a char boundary; it is inside '\u{a0}' (bytes 0..2) of ` data:,a` (thread Script(1,1), at components/script/dom/htmlimageelement.rs:1896)
 0:07.14 pid:15614   14: core::str::traits::<impl core::slice::index::SliceIndex<str> for core::ops::range::Range<usize>>::index
 0:07.14 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/str/traits.rs:241:21
 0:07.14 pid:15614   15: core::str::traits::<impl core::ops::index::Index<I> for str>::index
 0:07.14 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/str/traits.rs:61:9
 0:07.18 pid:15614   16: script::dom::htmlimageelement::collect_sequence_characters
 0:07.18 pid:15614              at /Users/jdm/src/servo/components/script/dom/htmlimageelement.rs:1896:23
 0:07.22 pid:15614   17: script::dom::htmlimageelement::parse_a_srcset_attribute
 0:07.22 pid:15614              at /Users/jdm/src/servo/components/script/dom/htmlimageelement.rs:1911:13
 0:07.22 pid:15614   18: script::dom::htmlimageelement::HTMLImageElement::update_source_set
 0:07.22 pid:15614              at /Users/jdm/src/servo/components/script/dom/htmlimageelement.rs:586:48
 0:07.22 pid:15614   19: script::dom::htmlimageelement::HTMLImageElement::select_image_source
 0:07.22 pid:15614              at /Users/jdm/src/servo/components/script/dom/htmlimageelement.rs:745:9
 0:07.22 pid:15614   20: script::dom::htmlimageelement::HTMLImageElement::update_the_image_data_sync_steps
 0:07.22 pid:15614              at /Users/jdm/src/servo/components/script/dom/htmlimageelement.rs:871:42
 0:07.22 pid:15614   21: <script::dom::htmlimageelement::ImageElementMicrotask as script::microtask::MicrotaskRunnable>::handler
 0:07.22 pid:15614              at /Users/jdm/src/servo/components/script/dom/htmlimageelement.rs:1427:21
 0:07.26 pid:15614   22: script::microtask::MicrotaskQueue::checkpoint
 0:07.26 pid:15614              at /Users/jdm/src/servo/components/script/microtask.rs:134:25
 0:07.29 pid:15614   23: script::script_thread::ScriptThread::perform_a_microtask_checkpoint
 0:07.29 pid:15614              at /Users/jdm/src/servo/components/script/script_thread.rs:4159:13
 0:07.29 pid:15614   24: script::script_thread::ScriptThread::handle_msgs
 0:07.29 pid:15614              at /Users/jdm/src/servo/components/script/script_thread.rs:1973:13
 0:07.29 pid:15614   25: script::script_thread::ScriptThread::start
 0:07.29 pid:15614              at /Users/jdm/src/servo/components/script/script_thread.rs:1454:15
 0:07.32 pid:15614   26: <script::script_thread::ScriptThread as script_layout_interface::ScriptThreadFactory>::create::{{closure}}::{{closure}}
 0:07.32 pid:15614              at /Users/jdm/src/servo/components/script/script_thread.rs:845:25
 0:07.35 pid:15614   27: profile_traits::mem::ProfilerChan::run_with_memory_reporting
 0:07.35 pid:15614              at /Users/jdm/src/servo/components/shared/profile/mem.rs:91:9
 0:07.35 pid:15614   28: <script::script_thread::ScriptThread as script_layout_interface::ScriptThreadFactory>::create::{{closure}}
 0:07.35 pid:15614              at /Users/jdm/src/servo/components/script/script_thread.rs:843:17
 0:07.35 pid:15614   29: std::sys_common::backtrace::__rust_begin_short_backtrace
 0:07.35 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/sys_common/backtrace.rs:155:18
 0:07.38 pid:15614   30: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
 0:07.38 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/thread/mod.rs:528:17
 0:07.41 pid:15614   31: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
 0:07.41 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/panic/unwind_safe.rs:272:9
 0:07.44 pid:15614   32: std::panicking::try::do_call
 0:07.44 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panicking.rs:552:40
 0:07.44 pid:15614   33: ___rust_try
 0:07.44 pid:15614   34: std::panicking::try
 0:07.44 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panicking.rs:516:19
 0:07.44 pid:15614   35: std::panic::catch_unwind
 0:07.44 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panic.rs:146:14
 0:07.44 pid:15614       std::thread::Builder::spawn_unchecked_::{{closure}}
 0:07.44 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/thread/mod.rs:527:30
 0:07.82 pid:15614   36: core::ops::function::FnOnce::call_once{{vtable.shim}}
 0:07.82 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/ops/function.rs:250:5
 0:07.82 pid:15614   37: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
 0:07.82 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/boxed.rs:2020:9
 0:07.82 pid:15614       <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
 0:07.82 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/alloc/src/boxed.rs:2020:9
 0:07.82 pid:15614       std::sys::pal::unix::thread::Thread::new::thread_start
 0:07.82 pid:15614              at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/sys/pal/unix/thread.rs:108:17
 0:07.82 pid:15614   38: __pthread_joiner_wake
```

### PR DESCRIPTION
**Pull request by Taym95**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #32976

- [x] There are tests for these changes OR

### CODE REVIEW

File: components/script/dom/htmlimageelement.rs [hunk 1]
```diff
@@ -1891,7 +1891,8 @@ pub fn collect_sequence_characters<F>(s: &str, predicate: F) -> (&str, &str)
 where
     F: Fn(&char) -> bool,
 {
-    for (i, ch) in s.chars().enumerate() {
+    let mut char_indices = s.char_indices();
+    while let Some((i, ch)) = char_indices.next() {
```
Comment (mrobinson) [inline components/script/dom/htmlimageelement.rs:1895]:
```suggestion
    for (index, character) in s.char_indices() {
```

**Review by mrobinson — CHANGES_REQUESTED** [review]
Nice fix! I have one small suggestion:

**Comment by NotnaKO** [comment]
Why don't we use rfind here? Like:

**Comment by NotnaKO** [comment]
Why don't we use find here?
```rs
let i = s.find(|c| !predicate(c)).unwrap_or(s.len())
(&s[0..i], &s[i..])
```

**Comment by Taym95** [comment]
> Why don't we use find here? Like:
> 
> ```rust
> let i = s.find(|c| !predicate(&c)).unwrap_or(s.len());
> (&s[0..i], &s[i..])
> ```
> 
> PR: #32980

Will give same result, I am trying to debug this one

**Pull request closed without merging**